WalletWasabi/Affiliation/AffiliateServerStatusUpdater.cs

```diff
@@ -0,0 +1,96 @@
+using System.Linq;
+using System.Collections.Immutable;
+using System.Collections.Generic;
+using WalletWasabi.Affiliation.Models;
+using System.Threading;
+using System.Threading.Tasks;
+using WalletWasabi.Bases;
+
+namespace WalletWasabi.Affiliation;
+
+public class AffiliateServerStatusUpdater : PeriodicRunner
+{
+	private static TimeSpan AffiliateServerTimeout = TimeSpan.FromSeconds(20);
+	private static TimeSpan Interval = TimeSpan.FromMinutes(1);
+
+	public AffiliateServerStatusUpdater(IDictionary<AffiliationFlag, AffiliateServerHttpApiClient> clients)
+		  : base(Interval)
+	{
+		Clients = clients;
+		RunningAffiliateServers = ImmutableList<AffiliationFlag>.Empty;
+	}
+
+	private IDictionary<AffiliationFlag, AffiliateServerHttpApiClient> Clients { get; }
+	public ImmutableList<AffiliationFlag> RunningAffiliateServers { get; private set; }
+
+	public IEnumerable<AffiliationFlag> GetRunningAffiliateServers()
+	{
+		return RunningAffiliateServers;
+	}
+
+	protected override async Task ActionAsync(CancellationToken cancel)
+	{
+		await UpdateRunningAffiliateServersAsync();
+	}
+
+	private static async Task<bool> IsAffiliateServerRunning(AffiliateServerHttpApiClient client, CancellationToken cancellationToken)
+	{
+		try
+		{
+			StatusResponse result = await client.GetStatus(new StatusRequest(), cancellationToken);
+			return true;
+		}
+		catch (Exception exception)
```

Comment:
> Catch only the exception's you are interested on.

I'm interested in this exception. The purpose of this function is nothing else than to determine whether an affiliate is responding.

Reply:
In that case catch the appropriated exception.

Reply:
Is this what you mean? https://github.com/zkSNACKs/WalletWasabi/pull/9769/commits/c0956ea60df9215b44e33d91051c15aa51a12199

To be honest, it should be in your interest not to publish the affiliation flag if anything goes wrong.

Reply:
I reverted https://github.com/zkSNACKs/WalletWasabi/commit/c0956ea60df9215b44e33d91051c15aa51a12199. I still think that it is not completely wrong to catch _all_ exceptions because it is safer not to publish the affiliation flag if _anything_ fails.

> AffiliateServerStatusUpdater.IsAffiliateServerRunningAsync implementation would be much simpler with a for-loop + using an AffiliateServerHttpApiClient.SendAsync overload that is dumb and does not support requestNumber.

There is a functionality that repeats http request if the previous request didn't succeed. If I understand it right, you want to move this functionality from from `AffiliateServerHttpApiClient` to `AffiliateServerStatusUpdater`. I don't think that is right, because this functionality has nothing to do with the `AffiliateServerStatusUpdater`. `AffiliateServerStatusUpdater` should be agnostic to the implementation of `AffiliateServerHttpApiClient`.